File: web/api/price/[ticker].py
"""
GET /api/price/<ticker> — 실시간 주가 + ATR(14) + 52주 고저 + 시총

설계:
- yfinance 단독 사용 (Vercel us-region에서 KRX 직접 접근 불가 → 한국 종목은 .KS/.KQ 자동 시도)
- in-memory TTL 캐시 (5분) — 동일 인스턴스 내 hit 시 ~50ms
- Rate limit 60/분 per-IP (단순 sliding window, in-memory)
- CORS: 자체 origin + localhost:4321 (dev)
- graceful: 데이터 없거나 yfinance 오류 시 JSON {"error": ...} 200 (클라이언트가 처리)

매개변수:
- path 변수 ticker (Vercel `[ticker].py` 자동 매핑)
- 또는 ?ticker= query (보조)
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json
import os
import time
from datetime import datetime, timedelta

# ─────────────────────────────────────────────────────────────────────────
# CORS allow-list
# ─────────────────────────────────────────────────────────────────────────
ALLOWED_ORIGINS = {
    'https://stock-analyst-jungwon1.vercel.app',
    'http://localhost:4321',
    'http://127.0.0.1:4321',
}

# ─────────────────────────────────────────────────────────────────────────
# In-memory TTL cache (per warm instance)
# ─────────────────────────────────────────────────────────────────────────
_CACHE_TTL = 300  # 5분
_cache: dict[str, tuple[float, dict]] = {}

def _cache_get(key: str):
    rec = _cache.get(key)
    if not rec:
        return None
    ts, data = rec
    if time.time() - ts > _CACHE_TTL:
        _cache.pop(key, None)
        return None
    return data

def _cache_set(key: str, data: dict):
    _cache[key] = (time.time(), data)

# ─────────────────────────────────────────────────────────────────────────
# Rate limit (60 req/min per IP, sliding window)
# ─────────────────────────────────────────────────────────────────────────
_RATE_LIMIT = 60
_RATE_WINDOW = 60.0
_rate: dict[str, list[float]] = {}

def _rate_check(ip: str) -> bool:
    now = time.time()
    history = _rate.get(ip, [])
    history = [t for t in history if now - t < _RATE_WINDOW]
    if len(history) >= _RATE_LIMIT:
        return False
    history.append(now)
    _rate[ip] = history
    return True

# ─────────────────────────────────────────────────────────────────────────
# Ticker → yfinance symbol resolver
# ─────────────────────────────────────────────────────────────────────────
def _resolve_symbols(ticker: str) -> list[str]:
    """티커를 yfinance 심볼 후보 목록으로 변환. 한국 6자리 → .KS, .KQ 시도."""
    t = ticker.strip().upper()
    if not t:
        return []
    if t.isdigit() and len(t) == 6:
        return [f'{t}.KS', f'{t}.KQ']
    return [t]

# ─────────────────────────────────────────────────────────────────────────
# ATR(14) 계산
# ─────────────────────────────────────────────────────────────────────────
def _calc_atr(hist, period: int = 14):
    if len(hist) < period + 1:
        return None
    highs = hist['High'].values
    lows = hist['Low'].values
    closes = hist['Close'].values
    trs = []
    for i in range(1, len(hist)):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    if len(trs) < period:
        return None
    return sum(trs[-period:]) / period

# ─────────────────────────────────────────────────────────────────────────
# 핵심: yfinance fetch
# ─────────────────────────────────────────────────────────────────────────
def fetch_price(ticker: str) -> dict:
    import yfinance as yf

    symbols = _resolve_symbols(ticker)
    if not symbols:
        return {'error': 'ticker 비어있음', 'ticker': ticker}

    last_err = None
    for sym in symbols:
        try:
            stock = yf.Ticker(sym)
            hist = stock.history(period='3mo')
            if hist.empty:
                last_err = f'{sym}: 데이터 없음'
                continue

            hist = hist.tail(30)
            latest = hist.iloc[-1]
            current_price = round(float(latest['Close']), 4)

            atr = _calc_atr(hist, period=14)
            info = {}
            try:
                info = stock.info or {}
            except Exception:
                info = {}

            currency = info.get('currency', 'USD')
            name = info.get('shortName') or info.get('longName') or ticker
            market = info.get('exchange') or 'US'

            hist_52w = stock.history(period='1y')
            high_52w = round(float(hist_52w['High'].max()), 4) if not hist_52w.empty else None
            low_52w = round(float(hist_52w['Low'].min()), 4) if not hist_52w.empty else None

            prev_close = round(float(hist.iloc[-2]['Close']), 4) if len(hist) >= 2 else None
            change_pct = round((current_price / prev_close - 1) * 100, 3) if prev_close else None

            decimals = 0 if currency == 'KRW' else 2
            return {
                'ticker': ticker,
                'symbol': sym,
                'name': name,
                'market': market,
                'currency': currency,
                'current_price': round(current_price, decimals),
                'prev_close': round(prev_close, decimals) if prev_close else None,
                'change_pct': change_pct,
                'high_52w': round(high_52w, decimals) if high_52w else None,
                'low_52w': round(low_52w, decimals) if low_52w else None,
                'market_cap': info.get('marketCap'),
                'volume': int(latest['Volume']) if latest.get('Volume') else None,
                'atr_14': round(atr, decimals) if atr else None,
                'atr_pct': round(atr / current_price * 100, 3) if atr else None,
                'stop_loss_2atr': round(current_price - 2 * atr, decimals) if atr else None,
                'target_3atr': round(current_price + 3 * atr, decimals) if atr else None,
                'date': hist.index[-1].strftime('%Y-%m-%d'),
                'fetch_time': datetime.utcnow().isoformat() + 'Z',
                'cache_ttl': _CACHE_TTL,
            }
        except Exception as e:
            last_err = f'{sym}: {e}'
            continue

    return {'error': last_err or '데이터 없음', 'ticker': ticker, 'tried': symbols}

# ─────────────────────────────────────────────────────────────────────────
# Vercel handler
# ─────────────────────────────────────────────────────────────────────────
class handler(BaseHTTPRequestHandler):
    def _send_json(self, status: int, body: dict, origin: str = ''):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Cache-Control', f'public, s-maxage={_CACHE_TTL}, stale-while-revalidate=60')
        if origin and origin in ALLOWED_ORIGINS:
            self.send_header('Access-Control-Allow-Origin', origin)
            self.send_header('Vary', 'Origin')
        self.end_headers()
        self.wfile.write(json.dumps(body, ensure_ascii=False).encode('utf-8'))

    def do_OPTIONS(self):
        origin = self.headers.get('Origin', '')
        self.send_response(204)
        if origin in ALLOWED_ORIGINS:
            self.send_header('Access-Control-Allow-Origin', origin)
            self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.send_header('Vary', 'Origin')
        self.end_headers()

    def do_GET(self):
        origin = self.headers.get('Origin', '')

        # Rate limit
        ip = self.headers.get('x-forwarded-for', self.client_address[0]).split(',')[0].strip()
        if not _rate_check(ip):
            self._send_json(429, {'error': 'rate limit 60/min 초과'}, origin)
            return

        # ticker 추출 (path 변수 → query 보조)
        parsed = urlparse(self.path)
        # /api/price/AAPL → 마지막 segment
        path_segs = [s for s in parsed.path.split('/') if s]
        ticker = path_segs[-1] if path_segs else ''
        # path가 [ticker] 그대로 오는 경우 query fallback
        if ticker.startswith('[') or not ticker or ticker == 'price':
            qs = parse_qs(parsed.query)
            ticker = (qs.get('ticker') or [''])[0]

        if not ticker:
            self._send_json(400, {'error': 'ticker 미지정'}, origin)
            return

        # 캐시 hit
        cache_key = ticker.upper()
        cached = _cache_get(cache_key)
        if cached:
            self._send_json(200, {**cached, 'cached': True}, origin)
            return

        # fetch
        try:
            data = fetch_price(ticker)
            if 'error' not in data:
                _cache_set(cache_key, data)
            self._send_json(200, {**data, 'cached': False}, origin)
        except Exception as e:
            self._send_json(500, {'error': f'서버 오류: {e}', 'ticker': ticker}, origin)

File: api/price/[ticker].js
/**
 * GET /api/price/<ticker> — 실시간 주가 + ATR(14) + 52주 고저 + 시총
 *
 * [v3.41, 2026-08-25] Python(yfinance) → Node 재작성.
 * 배경: yfinance 의존성 + 프로젝트 파일 포함으로 함수 번들 273MB > 225MB 한도
 * → 8/13부터 전 배포 Error (본서버 12일 동결 사고). Node 재작성으로 번들 수 KB.
 *
 * 계약 (구 Python 함수와 동일 — 소비처 DailyPick/compare/timemachine 무수정):
 *   현재가·전일비·ATR14·손절/목표(2·3ATR)·52주 고저·시총·?at= 시간머신·KST fetch_time
 *   한국 6자리 → .KS/.KQ 자동 시도. 오류는 JSON {error} 200 (클라이언트 graceful).
 * 데이터 소스: Yahoo v8 chart API 직접 호출 (yfinance 와 동일 원천, SDK 없이).
 */

const CACHE_TTL = 300; // 5분 (CDN s-maxage 겸용)
const KST_OFFSET_MIN = 9 * 60;

const ALLOWED_ORIGINS = new Set([
  "https://stock-analyst-jungwon1.vercel.app",
  "http://localhost:4321",
  "http://127.0.0.1:4321",
]);

// warm 인스턴스 캐시 + 레이트리밋 (CDN 캐시가 1차 방어, 이건 보조)
const cache = new Map(); // key → {ts, data}
const rate = new Map(); // ip → [timestamps]
const RATE_LIMIT = 60;
const RATE_WINDOW_MS = 60_000;

function rateCheck(ip) {
  const now = Date.now();
  const hist = (rate.get(ip) || []).filter((t) => now - t < RATE_WINDOW_MS);
  if (hist.length >= RATE_LIMIT) return false;
  hist.push(now);
  rate.set(ip, hist);
  return true;
}

function resolveSymbols(ticker) {
  const t = String(ticker || "")
    .trim()
    .toUpperCase();
  if (!t) return [];
  if (/^\d{6}$/.test(t)) return [`${t}.KS`, `${t}.KQ`];
  return [t];
}

async function fetchJson(url, timeoutMs = 8000) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const r = await fetch(url, {
      signal: ctrl.signal,
      headers: {
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)",
      },
    });
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    return await r.json();
  } finally {
    clearTimeout(timer);
  }
}

/** Yahoo v8 chart → {bars: [{t,h,l,c,v}], meta} (null 바 제거) */
async function fetchChart(sym, params) {
  const qs = new URLSearchParams({ interval: "1d", ...params }).toString();
  const j = await fetchJson(
    `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(sym)}?${qs}`,
  );
  const result = j?.chart?.result?.[0];
  if (!result) throw new Error(j?.chart?.error?.description || "데이터 없음");
  const ts = result.timestamp || [];
  const q = result.indicators?.quote?.[0] || {};
  const bars = [];
  for (let i = 0; i < ts.length; i++) {
    const c = q.close?.[i];
    if (c == null) continue;
    bars.push({
      t: ts[i],
      h: q.high?.[i] ?? c,
      l: q.low?.[i] ?? c,
      c,
      v: q.volume?.[i] ?? null,
    });
  }
  return { bars, meta: result.meta || {} };
}

function calcAtr(bars, period = 14) {
  if (bars.length < period + 1) return null;
  const trs = [];
  for (let i = 1; i < bars.length; i++) {
    trs.push(
      Math.max(
        bars[i].h - bars[i].l,
        Math.abs(bars[i].h - bars[i - 1].c),
        Math.abs(bars[i].l - bars[i - 1].c),
      ),
    );
  }
  const tail = trs.slice(-period);
  return tail.reduce((a, b) => a + b, 0) / period;
}

function kstIso() {
  const d = new Date(Date.now() + KST_OFFSET_MIN * 60_000);
  return d.toISOString().replace(/\.\d{3}Z$/, "+09:00");
}

function rnd(v, decimals) {
  if (v == null || !Number.isFinite(v)) return null;
  const f = 10 ** decimals;
  return Math.round(v * f) / f;
}

// Yahoo crumb 인증 (quoteSummary 용) — yfinance 와 동일 흐름, warm 인스턴스 1시간 캐시
let crumbCache = null; // {cookie, crumb, ts}
const UA = { 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)' };

async function getCrumb() {
  if (crumbCache && Date.now() - crumbCache.ts < 3600_000) return crumbCache;
  const r1 = await fetch('https://fc.yahoo.com', { headers: UA, redirect: 'manual' });
  const cookie = (r1.headers.get('set-cookie') || '').split(';')[0];
  if (!cookie) throw new Error('쿠키 없음');
  const r2 = await fetch('https://query1.finance.yahoo.com/v1/test/getcrumb', {
    headers: { ...UA, Cookie: cookie },
  });
  const crumb = (await r2.text()).trim();
  if (!crumb || crumb.includes('<')) throw new Error('crumb 획득 실패');
  crumbCache = { cookie, crumb, ts: Date.now() };
  return crumbCache;
}

async function fetchMarketCap(sym) {
  // 실패 시 null — compare 페이지 graceful (구 Python 함수도 info 실패 시 동일)
  try {
    const { cookie, crumb } = await getCrumb();
    const url = `https://query2.finance.yahoo.com/v10/finance/quoteSummary/${encodeURIComponent(sym)}?modules=price&crumb=${encodeURIComponent(crumb)}`;
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 5000);
    try {
      const r = await fetch(url, { headers: { ...UA, Cookie: cookie }, signal: ctrl.signal });
      const j = await r.json();
      return j?.quoteSummary?.result?.[0]?.price?.marketCap?.raw ?? null;
    } finally {
      clearTimeout(timer);
    }
  } catch {
    crumbCache = null; // 만료 의심 — 다음 호출에서 재발급
    return null;
  }
}

async function fetchPrice(ticker, at) {
  const symbols = resolveSymbols(ticker);
  if (!symbols.length) return { error: "ticker 비어있음", ticker };

  let lastErr = null;
  for (const sym of symbols) {
    try {
      const { bars, meta } = await fetchChart(sym, { range: "1y" });
      if (!bars.length) {
        lastErr = `${sym}: 데이터 없음`;
        continue;
      }

      const recent = bars.slice(-30);
      const latest = recent[recent.length - 1];
      const currentPrice = latest.c;
      const atr = calcAtr(recent, 14);

      const currency = meta.currency || "USD";
      const decimals = currency === "KRW" ? 0 : 2;
      const name = meta.shortName || meta.longName || ticker;
      const market = meta.exchangeName || meta.fullExchangeName || "US";

      const high52 = Math.max(...bars.map((b) => b.h));
      const low52 = Math.min(...bars.map((b) => b.l));
      const prevClose = recent.length >= 2 ? recent[recent.length - 2].c : null;

      // ?at= 시간 머신 — 해당일 종가 + 이후 수익률
      let atPrice = null;
      let returnSinceAt = null;
      if (at && /^\d{4}-\d{2}-\d{2}$/.test(at)) {
        try {
          const p1 = Math.floor(new Date(`${at}T00:00:00Z`).getTime() / 1000);
          const { bars: atBars } = await fetchChart(sym, {
            period1: String(p1),
            period2: String(p1 + 10 * 86400),
          });
          if (atBars.length) {
            atPrice = atBars[0].c;
            returnSinceAt = rnd((currentPrice / atPrice - 1) * 100, 2);
          }
        } catch {
          /* at 실패는 무시 — 본 응답은 유효 */
        }
      }

      return {
        ticker,
        symbol: sym,
        name,
        market,
        currency,
        current_price: rnd(currentPrice, decimals),
        prev_close: prevClose != null ? rnd(prevClose, decimals) : null,
        change_pct: prevClose
          ? rnd((currentPrice / prevClose - 1) * 100, 3)
          : null,
        high_52w: rnd(high52, decimals),
        low_52w: rnd(low52, decimals),
        market_cap: await fetchMarketCap(sym),
        volume: latest.v != null ? Math.round(latest.v) : null,
        atr_14: atr != null ? rnd(atr, decimals) : null,
        atr_pct: atr != null ? rnd((atr / currentPrice) * 100, 3) : null,
        stop_loss_2atr:
          atr != null ? rnd(currentPrice - 2 * atr, decimals) : null,
        target_3atr: atr != null ? rnd(currentPrice + 3 * atr, decimals) : null,
        date: new Date(latest.t * 1000).toISOString().slice(0, 10),
        at: at || null,
        at_price: atPrice != null ? rnd(atPrice, decimals) : null,
        return_since_at_pct: returnSinceAt,
        fetch_time: kstIso(),
        fetch_tz: "Asia/Seoul",
        cache_ttl: CACHE_TTL,
      };
    } catch (e) {
      lastErr = `${sym}: ${e.message || e}`;
    }
  }
  return { error: lastErr || "데이터 없음", ticker, tried: symbols };
}

module.exports = async (req, res) => {
  const origin = req.headers.origin || "";
  const setCors = () => {
    if (ALLOWED_ORIGINS.has(origin)) {
      res.setHeader("Access-Control-Allow-Origin", origin);
      res.setHeader("Vary", "Origin");
    }
  };

  if (req.method === "OPTIONS") {
    setCors();
    res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");
    res.status(204).end();
    return;
  }

  setCors();
  res.setHeader("Content-Type", "application/json; charset=utf-8");
  res.setHeader(
    "Cache-Control",
    `public, s-maxage=${CACHE_TTL}, stale-while-revalidate=60`,
  );

  const ip =
    String(req.headers["x-forwarded-for"] || "")
      .split(",")[0]
      .trim() || "unknown";
  if (!rateCheck(ip)) {
    res.status(429).json({ error: "rate limit 60/min 초과" });
    return;
  }

  const ticker = String(req.query.ticker || "").trim();
  const at = String(req.query.at || "").trim();
  if (!ticker) {
    res.status(400).json({ error: "ticker 미지정" });
    return;
  }

  const cacheKey = `${ticker.toUpperCase()}|${at}`;
  const hit = cache.get(cacheKey);
  if (hit && Date.now() - hit.ts < CACHE_TTL * 1000) {
    res.status(200).json({ ...hit.data, cached: true });
    return;
  }

  try {
    const data = await fetchPrice(ticker, at);
    if (!data.error) cache.set(cacheKey, { ts: Date.now(), data });
    res.status(200).json({ ...data, cached: false });
  } catch (e) {
    res.status(500).json({ error: `서버 오류: ${e.message || e}`, ticker });
  }
};
